src/modules/navigator/MissionFeasibility/FeasibilityChecker.cpp

```diff
@@ -0,0 +1,692 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in
+ *    the documentation and/or other materials provided with the
+ *    distribution.
+ * 3. Neither the name PX4 nor the names of its contributors may be
+ *    used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ * FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ * COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ * BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
+ * OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
+ * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ *
+ ****************************************************************************/
+
+#include "FeasibilityChecker.hpp"
+#include <systemlib/mavlink_log.h>
+#include <px4_platform_common/events.h>
+#include <drivers/drv_pwm_output.h>
+#include "../mission_block.h"
+#include <lib/mathlib/mathlib.h>
+#include <lib/geo/geo.h>
+
+FeasibilityChecker::FeasibilityChecker() :
+	ModuleParams(nullptr)
+{
+}
+
+void FeasibilityChecker::reset()
+{
+	_mission_validity_failed = false;
+	_takeoff_failed = false;
+	_fixed_wing_landing_failed = false;
+	_distance_first_waypoint_failed = false;
+	_distance_between_waypoints_failed = false;
+	_below_home_alt_failed = false;
+	_fixed_wing_land_approach_failed = false;
+	_takeoff_land_available_failed = false;
+
+	_found_item_with_position = false;
+	_has_vtol_takeoff = false;
+	_has_takeoff = false;
+
+	_landing_valid = false;
+	_do_land_start_index = -1;
+	_landing_approach_index = -1;
+	_mission_item_previous = {};
+
+	_first_waypoint_found = false;
+
+	_last_lat = NAN;
+	_last_lon = NAN;
+	_last_cmd = -1;
+}
+
+void FeasibilityChecker::updateData()
+{
+	home_position_s home = {};
+
+	if (_home_pos_sub.copy(&home)) {
+		if (home.valid_hpos) {
+			_home_lat_lon = matrix::Vector2d(home.lat, home.lon);
+		}
+
+		if (home.valid_alt) {
+			_home_alt_msl = home.alt;
+		}
+	}
+
+	vehicle_status_s status = {};
+
+	if (_status_sub.copy(&status)) {
+		_vehicle_type = status.vehicle_type;
+		_is_vtol = status.is_vtol;
```

Comment:
Can _vehicle_type and _is_vtol be combined?

Reply:
@sfuhrer No, you would not be able to distinguish between a VTOL and a Multirotor.

Reply:
What I meant is that instead of _vehicle_type and _is_vtol, you could make en enum for the vehicle type, with the list being MC, FW and VTOL.

Reply:
addressed